### ISSUE TITLE
Ice trickle support

### ISSUE THREAD

**Issue by rprata** [aiortc/aiortc#691]
Hi again, 

I would to open again the discussion about ice tickle support. So, is it so difficult to implement in aiortc codebase? Could explain how can we implement this in `aiortc` (better way to implement)? I could help you in this issue.

**Comment by mglettig** [comment]
I would like to join this discussion. I would need ICE trickle support to get compatibility to my WebRTC counterpart that seems to only support offering with ICE trickle. When searching for ICE trickle support in this code base I found some hints that it should be possible to get it working but I didn't found out how. For example I found this discussion here very interesting https://github.com/aiortc/aiortc/issues/651 but it looks like the issue was not resolved for the one that raised the question.

**Comment by jlaine** [comment]
> I would like to join this discussion. I would need ICE trickle support to get compatibility to my WebRTC counterpart that seems to only support offering with ICE trickle. When searching for ICE trickle support in this code base I found some hints that it should be possible to get it working but I didn't found out how. For example I found this discussion here very interesting #651 but it looks like the issue was not resolved for the one that raised the question.

This is off topic, we support *receiving* trickled ICE candidates, so there should be no interop issue.

**Comment by jlaine** [comment]
@rprata this is where it needs to happen: https://github.com/aiortc/aioice/issues/4

**Comment by rprata** [comment]
> @rprata this is where it needs to happen: [aiortc/aioice#4](https://github.com/aiortc/aioice/issues/4)

I'll check this issue and update it. 

Thanks.

**Comment by jlaine** [comment]
Hi @rprata I'm happy to keep this issue open as there will be some integration work in aiortc itself too.

**Comment by mglettig** [comment]
> > I would like to join this discussion. I would need ICE trickle support to get compatibility to my WebRTC counterpart that seems to only support offering with ICE trickle. When searching for ICE trickle support in this code base I found some hints that it should be possible to get it working but I didn't found out how. For example I found this discussion here very interesting #651 but it looks like the issue was not resolved for the one that raised the question.
> 
> This is off topic, we support _receiving_ trickled ICE candidates, so there should be no interop issue.

I think there still may occur interop issues: That is if the counterpart of aiortc can't handle it when the ICE candidates are sent directly in the answer. But this could be more of an issue of the other library than an issue of aiortc...
If I understand you correctly aiortc supports _receiving_ ICE trickle candidates out of the box (no special handling needed in the application)? Is there any particular library that works particularly well as a counterpart of aiortc? I think the one from google <https://webrtc.googlesource.com/src> is particularly widespread. Was it tested with aiortc?

Sorry if this post is off topic but I didn't find a better place to post it.

**Comment by jlaine** [comment]
@mglettig we explicitly support both candidates passed into `setRemoteDescription` or subsequenty with `addIceCandidate`. If you have a reproducible scenario where this does not work, please open a separate issue.

**Comment by mglettig** [comment]
> @mglettig we explicitly support both candidates passed into `setRemoteDescription` or subsequenty with `addIceCandidate`. If you have a reproducible scenario where this does not work, please open a separate issue.

Just confirming that it works as you mentioned. 👍 The problem was on my side (my signalling adapter blocked a thread where aiortc was running it leading to no response during ICE exchange).

**Comment by abhinavsingh** [comment]
@jlaine We are trying both controlled and controlling scenario for our integration, but we are hitting a road blocker with both strategy:

1) When aiortc is the one sending an offer, DTLS handshake fails and no media gets transmitted.   This happens on Google Cloud (COS) VM running our Docker instance.   Locally, we weren't able to reproduce DTLS handshake issues (even when running within a Docker VM locally) when aiortc is the offerer.   We didn't catch this case until we deployed on cloud recently.

2) When aiortc is the one answering to remote offer, it doesn't include candidates in the answer.   As a result, we experience that remote client is able to receive aiortc audio, but aiortc is unable to receive audio from the remote client.  We are able to reproduce this both locally and on cloud deployed instances.

Do you have any guidance here, as to:

a) How can we ensure answers include candidate lines
b) OR iff you have any advice about DTLS handshake failures in Google Cloud environment, will be much appreciated.

Thank you

**Comment by guidocalvano** [comment]
> > This is off topic, we support receiving trickled ICE candidates, so there should be no interop issue.
> 
> In order to get around [Firefox's bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1691189) with resolving mDNS addresses only after `setRemoteCandidate` is called. I tried and seemingly succeeded to implement trickle candidate support. I updated the webcam.py example with another endpoint for receiving trickled candidates. ICE connections in Firefox are now successful.
> 
> However, despite `iceConnectionState` equaling `connected` I still cannot view webcam video in my application. Checking `about:webrtc` shows a successful connection but no data being transmitted. (However, I do see the feed if I turn on STUN locally to generate a srflx candidate and don't use trickling)
> 
> @jlaine Is it expected that I should be able to see the video feed after an ICE connection is made via trickling? Is there anything special I need to do when sending the video feed in this case?

Can you share your solution or @jlaine can a working Firefox example be supplied?
